### PR TITLE
solana/program_v2: Add withdraw function for gatekeeper

### PR DESCRIPTION
**Pull request by JulianSpring**
Added withdraw for the gatekeeper

### CODE REVIEW

File: solana/program_v2/packages/tests/src/pass-suite/issue.test.ts [hunk 1]
```diff
@@ -81,7 +81,8 @@ describe('issue', () => {
         funderKeypair.publicKey
       )
       .withPartialSigners(funderKeypair)
-      .rpc();
+      .rpc()
+      .catch((e) => console.log(e));
```
Comment (JulianSpring) [inline solana/program_v2/packages/tests/src/pass-suite/issue.test.ts:85]: Need to remove

File: solana/program_v2/packages/client/core/src/NetworkService.ts [hunk 2]
```diff
@@ -296,16 +296,24 @@ export class NetworkService extends AbstractService {
...
+    receiverTokenAccount: PublicKey,
+    gatekeeperTokenAccount: PublicKey,
+    amount: number
   ): ServiceBuilder {
     const instructionPromise = this._program.methods
-      .gatekeeperWithdraw()
+      .gatekeeperWithdraw(new anchor.BN(amount))
       .accounts({
-        gatekeeper: this._gatekeeperAccount,
+        gatekeeper: gatekeeper,
```
Comment (rado0x54) [inline solana/program_v2/packages/client/core/src/NetworkService.ts:310]: assignment can be removed

File: solana/program_v2/programs/gateway_v2/src/lib.rs [hunk 3]
```diff
@@ -61,8 +63,9 @@ pub mod solana_anchor_gateway {
         instructions::network::set_gatekeeper_state(ctx, state)
     }
 
-    pub fn gatekeeper_withdraw(ctx: Context<GatekeeperWithdrawAccount>) -> Result<()> {
-        instructions::network::gatekeeper_withdraw(ctx)
+    pub fn gatekeeper_withdraw(ctx: Context<GatekeeperWithdrawAccount>, amount: u64) -> Result<()> {
+        msg!("Withdrawing {} from gatekeeper", amount);
```
Comment (JulianSpring) [inline solana/program_v2/programs/gateway_v2/src/lib.rs:67]: Need to remove

File: solana/program_v2/packages/client/core/src/NetworkService.ts (hunk 2 above)
Comment (JulianSpring) [reply solana/program_v2/packages/client/core/src/NetworkService.ts:310]: Ah yeah thanks!

File: solana/program_v2/packages/tests/src/test-set-up.ts [hunk 4]
```diff
@@ -126,6 +128,15 @@ export const setUpAdminNetworkGatekeeper = async (
     .withPartialSigners(networkAuthority)
     .rpc();
 
+  const data = {
```
Comment (rado0x54) [inline solana/program_v2/packages/tests/src/test-set-up.ts:131]: See lint warning.

File: solana/program_v2/packages/tests/src/test-set-up.ts [hunk 5]
```diff
@@ -8,7 +8,9 @@ import {
 import {
   AdminService,
   airdrop,
+  CreateGatekeeperData,
   GatekeeperService,
+  GATEWAY_PROGRAM_ID,
```
Comment (rado0x54) [inline solana/program_v2/packages/tests/src/test-set-up.ts:13]: Can be removed (lint error)

File: solana/program_v2/programs/gateway_v2/src/instructions/network/gatekeeper_withdraw.rs [hunk 6]
```diff
@@ -1,31 +1,62 @@
...
     pub gatekeeper: Account<'info, Gatekeeper>,
-    #[account(mut)]
+    #[account()]
     pub authority: Signer<'info>,
-    /// CHECK: Receiver destination account does not need to satisfy the any constraints.
-    #[account(mut)]
-    pub receiver: UncheckedAccount<'info>,
     pub system_program: Program<'info, System>,
+    pub spl_token_program: Program<'info, Token>,
+    pub mint_account: Account<'info, Mint>,
```
Comment (rado0x54) [inline solana/program_v2/programs/gateway_v2/src/instructions/network/gatekeeper_withdraw.rs:57]: We should add a comment here that verification is done in CODE. @JulianSpring or do you think there's an elegant way to do it here?

Comment (JulianSpring) [reply solana/program_v2/programs/gateway_v2/src/instructions/network/gatekeeper_withdraw.rs:57]: I'll add a comment and think of a better solution! Thanks

File: solana/program_v2/packages/tests/src/test-set-up.ts [hunk 7]
```diff
@@ -8,6 +8,7 @@ import {
 import {
   AdminService,
   airdrop,
+  CreateGatekeeperData,
```
Comment (william-brooks) [inline solana/program_v2/packages/tests/src/test-set-up.ts:11]: Linting

**Review by william-brooks — APPROVED** [review]

**Pull request merged**